pkg/tbtc/inactivity.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		return fmt.Errorf("could not get wallet members info: [%v]", err)
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(ice.signers))
+
+	for _, currentSigner := range ice.signers {
+		ice.protocolLatch.Lock()
+		defer ice.protocolLatch.Unlock()
+
+		defer wg.Done()
+
+		go func(signer *signer) {
+			execLogger.Info(
+				"[member:%v] starting inactivity claim publishing",
+				signer.signingGroupMemberIndex,
+			)
+
+			ctx, cancelCtx := withCancelOnBlock(
+				context.Background(),
+				timeoutBlock,
+				ice.waitForBlockFn,
+			)
+			defer cancelCtx()
+
+			subscription := ice.chain.OnInactivityClaimed(
+				func(event *InactivityClaimedEvent) {
+					defer cancelCtx()
+
+					execLogger.Infof(
+						"[member:%v] Inactivity claim submitted for wallet "+
+							"with ID [0x%x] and nonce [%v] by notifier [%v] "+
+							"at block [%v]",
+						signer.signingGroupMemberIndex,
+						event.WalletID,
+						event.Nonce,
+						event.Notifier,
+						event.BlockNumber,
+					)
+				})
+			defer subscription.Unsubscribe()
+
+			err := ice.publish(
+				ctx,
+				execLogger,
+				sessionId,
+				signer.signingGroupMemberIndex,
+				wallet.groupSize(),
+				wallet.groupDishonestThreshold(
+					ice.groupParameters.HonestThreshold,
+				),
+				groupMembers,
+				ice.membershipValidator,
+				claim,
+			)
+			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					execLogger.Infof(
+						"[member:%v] inactivity claim is no longer awaiting "+
+							"publishing; aborting inactivity claim publishing",
+						signer.signingGroupMemberIndex,
+					)
+					return
+				}
+
+				execLogger.Errorf(
+					"[member:%v] inactivity claim publishing failed [%v]",
+					signer.signingGroupMemberIndex,
+					err,
+				)
+				return
+			}
+		}(currentSigner)
+	}
+
+	// Wait until all controlled signers complete their routine.
+	wg.Wait()
+
+	return nil
+}
+
+func (ice *inactivityClaimExecutor) getWalletOperatorsIDs() ([]uint32, error) {
+	// Cache mapping operator addresses to their wallet member IDs. It helps to
+	// limit the number of calls to the ETH client if some operator addresses
+	// occur on the list multiple times.
+	operatorIDCache := make(map[chain.Address]uint32)
+
+	walletMemberIDs := make([]uint32, 0)
+
+	for _, operatorAddress := range ice.wallet().signingGroupOperators {
+		// Search for the operator address in the cache. Store the operator
+		// address in the cache if it's not there.
+		if operatorID, found := operatorIDCache[operatorAddress]; !found {
+			fetchedOperatorID, err := ice.chain.GetOperatorID(operatorAddress)
+			if err != nil {
+				return nil, fmt.Errorf("could not get operator ID: [%w]", err)
+			}
+			operatorIDCache[operatorAddress] = fetchedOperatorID
+			walletMemberIDs = append(walletMemberIDs, fetchedOperatorID)
+		} else {
+			walletMemberIDs = append(walletMemberIDs, operatorID)
+		}
+	}
+
+	return walletMemberIDs, nil
+}
+
+func (ice *inactivityClaimExecutor) publish(
```

Comment:
Right now the `inactivityClaimExecutor` exposes `publishClaim` and `publish` functions. This is ambiguous for external clients. Maybe we can rename them as follows:
- `publishClaim` -> `claimInactivity`. This will follow the pattern of all other executors, e.g. `signingExecutor.sign` or `coordinationExecutor.coordinate`
- `publish` -> `publishInactivityClaim`. Here we go down into the details of "claiming inactivity". Within this function, we execute `inactivity.PublishClaim` which triggers the protocol.

Reply:
Done in b2fcba0ad3fd07397312b04c820de1c7dcdf7daf.